lib/Sema/CSOptimizer.cpp

```diff
@@ -281,6 +470,52 @@ static void findFavoredChoicesBasedOnArity(
     favoredChoice(choice);
 }
 
+/// Preserves old behavior where, for unary calls, the solver would not previously
+/// consider choices that didn't match on the number of parameters (regardless of
+/// defaults and variadics) and only exact matches were favored.
+static std::optional<DisjunctionInfo> preserveFavoringOfUnlabeledUnaryArgument(
+    ConstraintSystem &cs, Constraint *disjunction, ArgumentList *argumentList) {
+  if (!argumentList->isUnlabeledUnary())
+    return std::nullopt;
+
+  auto ODRE = isOverloadedDeclRef(disjunction);
+  bool preserveFavoringOfUnlabeledUnaryArgument =
+      !ODRE || numOverloadChoicesMatchingOnArity(ODRE, argumentList) < 2;
+
+  if (!preserveFavoringOfUnlabeledUnaryArgument)
+    return std::nullopt;
+
+  auto *argument =
+      argumentList->getUnlabeledUnaryExpr()->getSemanticsProvidingExpr();
+  // The hack operated on "favored" types and only declaration references,
+  // applications, and (dynamic) subscripts had them if they managed to
+  // get an overload choice selected during constraint generation.
+  if (!(isExpr<DeclRefExpr>(argument) || isExpr<ApplyExpr>(argument) ||
+        isExpr<SubscriptExpr>(argument) ||
+        isExpr<DynamicSubscriptExpr>(argument)))
+    return {/*score=*/0};
+
+  auto argumentType = cs.getType(argument);
+  if (argumentType->hasTypeVariable() || argumentType->hasDependentMember())
```

Comment:
Is it possible for hasDependentMember() to return true while hasTypeVariable() returns false?

Reply:
Yeah, solver would form dependent member types when base is invalid as well